File: data/graphs/csr.py
from time import time
from random import randrange,seed

"""	Compressed sparse row graph class
"""
class CSRGraph:
	def __init__(self):
		self.num_verts = 0
		self.num_edges = 0
		self.num_rows = 0
		self.vals = []
		self.cols = []
		self.rows = []

	"""	Insert a directed edge between vertices v1 and v2. Bidirectional not implemented yet.
	"""
	def insertEdge(self, v1: int, v2: int, weight: float):
		# Extend if necessary
		largest_vert = max(v1, v2)
		if largest_vert >= self.num_rows:
			extend = [self.num_edges for i in range(largest_vert + 2 - self.num_rows)]
			# If empty, can't remove last item
			if len(self.rows) > 0:
				self.rows.pop()
			self.rows.extend(extend)

			self.num_rows = largest_vert+1
			self.rows[largest_vert+1] = self.num_edges

		# Find index to insert
		colIdx = self.rows[v1]
		nextIdx = self.rows[v1+1]
		i = colIdx
		row = self.cols[colIdx:nextIdx]
		if v2 in row:
			return False

		self.vals.insert(i, weight)
		self.cols.insert(i, v2)
		self.num_edges += 1

		for i in range(v1+1, self.num_rows + 1):
			self.rows[i] += 1

		return True

	"""	Returns a pointer to the location in the graph array where the neighbors of
		 v1 start. The ret_size parameter will be set to the count of v1 neighbors.
		If an invalid parameter is given, ret_size will remain untouched, and the
		 function will return NULL.
	"""
	def getNeighbors(self, v1: int) -> list:
		if v1 >= self.num_rows:
			return []

		return self.vals[self.rows[v1]:self.rows[v1+1]]

	"""	Print out all 3 arrays in the graph
	"""
	def __str__(self):
		ret = ",".join([str(a) for a in self.vals]) + "\n"
		ret += ",".join([str(a) for a in self.cols]) + "\n"
		ret += ",".join([str(a) for a in self.rows])
		return ret


def checkSizes(graph, insert_count: list, N, offset):
	print("Checking row size and columns..")

	max_count = 0
	max_vert_diff = N - graph.num_rows + offset
	#print("Max diff:", max_vert_diff,"\tNum rows:", graph.num_rows, "\tN:", N)
	# For each row
	for i in range(N - max_vert_diff):
		if insert_count[i] > max_count:
			max_count = insert_count[i]

		# Check size
		row_size = graph.rows[i+1 + offset] - graph.rows[i + offset]
		if row_size != insert_count[i]:
			print("SIZE MISMATCH FOR ROW %u! Got: %u\texpected: %u"%(i, row_size, insert_count[i]))
			print("INDICES %2u and %2u"%(i+1+offset, i+offset))
	print("Check done. Max count: %u"%max_count)

	# Check NNZ entry at end of row array
	idx = N + offset - max_vert_diff
	if graph.rows[idx] != N:
		print("SIZE MISMATCH! Got: %u\t(%u)\texpected: %u\n"%(graph.rows[idx], graph.rows[N + offset], N))

def checkData(graph, insert_count: list, edge_check, N, offset):
	max_vert_diff = N - graph.num_rows + offset
	print("Num_rows: %u\tdiff: %u\n" % (graph.num_rows, max_vert_diff))

	for i in range(N - max_vert_diff):
		# Check data validity
		row_start = graph.rows[i + offset]

		# For each column
		for j in range(insert_count[i]):
			idx = row_start + j
			col = graph.cols[idx] - offset

			if graph.vals[idx] != edge_check[i][col]:
				print("VAL MISMATCH FOR (%u, %u). Got: %u\texpected: %u"%(i, col, graph.vals[idx], edge_check[i][col]))

def main():
	seed(time())

	gr = CSRGraph()

	N = 5000

	# Will be dense matrix that stores same information as CSR
	edge_check = [[0.0]*N for i in range(N)]

	# Count the size of each row during insert
	simple_count = [0] * N

	# Insert random vertex pairings, in range [offset, N + offset)
	offset = 1
	dup_count = 0
	i = 0
	while i < N:
		a = randrange(N) + offset
		b = randrange(N) + offset

		#print("Inserting (%d, %d)"%(a, b))
		ret = gr.insertEdge(a, b, a*b/2.0)
		#print(gr)
		if not ret:
			dup_count += 1
			continue

		# Truth data
		edge_check[a-offset][b-offset] = a*b/2.0
		simple_count[a - offset] += 1

		i += 1
	print("%u duplicates encountered.\n" % dup_count)

	#print(gr)

	#	Make sure that each row is the correct size, and that the row has the
	#	 correct data in each column.
	checkSizes(gr, simple_count, N, offset)
	#print("\n".join([str(a) for a in edge_check]))
	checkData(gr, simple_count, edge_check, N, offset)

	return True

main()

File: data/graphs/csr.c
/***	Author: Chris Skane
		Created: 26 MAR 2020
		Description: Graph structure using the Compressed-Sparse-Row format
		Notes:
		- An edge with no vertices might be an edge with a self-vertex of weight 0 (or Inf or NaN).
		- An edge with no vertices might also not be allowed (CSR limitation)

		TODO: Add vertices/getVertices function
		TODO: Add updateEdge/updateWeight function
***/

#include"csr.h"

/*	Add to num_vert when allocating, simply a way to offset initial allocation
	 size without largely affecting
*/
#define NUM_EDGE_OFFSET 255

#define MAX(x,y) ((x) > (y) ? (x) : (y))

/**	Zeroes all struct entries
**/
void initGraph(struct csr_graph *graph){
	if(graph == NULL) return;
	graph->num_edges = 0;
	graph->num_verts = 0;
	graph->num_rows = 0;
	graph->vals = NULL;
	graph->cols = NULL;
	graph->rows = NULL;
}

void destroyGraph(struct csr_graph *graph){
	if(graph == NULL) return;
	graph->num_edges = 0;
	graph->num_verts = 0;
	graph->num_rows = 0;
	free(graph->vals);
	free(graph->cols);
	free(graph->rows);
}

long shiftUp(unsigned int *arr, size_t size, size_t offset){
	if(offset < 0 || size < 1) return 0;

	size_t i = size+1;
	size_t count = 1;
	while(--i > offset && count <= size){
		arr[i] = arr[size-count];
		++count;
	}
}

/**	Insert a directed edge between vertices v1 and v2. Bidirectional not implemented yet.
**/
long insertArc(struct csr_graph *graph, unsigned int v1, unsigned int v2, double weight){
	if(graph == NULL) return -1;

	// Check if the array needs to be allocated. Requires that EDGE_OFFSET is of form 2^n - 1
	int num_leads = __builtin_clzl(graph->num_edges + NUM_EDGE_OFFSET); // Returns 31/63 for x == 0
	int new_num_leads = __builtin_clzl(graph->num_edges + NUM_EDGE_OFFSET + 1);

	if(num_leads > new_num_leads){
		size_t alloc_size = (1 << (64 - new_num_leads)) * sizeof(*graph->vals);
		//printf("alloc_size: %lu\tnum_edge: %lu\n", alloc_size, graph->num_edges);
		double *vals_temp = realloc(graph->vals, alloc_size);
		if(vals_temp == NULL){
			fprintf(stderr, "Could not allocate more space for vertices\n");
			return -1;
		}
		graph->vals = vals_temp;

		alloc_size = (1 << (64 - new_num_leads)) * sizeof(*graph->cols);
		//printf("alloc_size2: %lu\n", alloc_size);
		unsigned int *cols_temp = realloc(graph->cols, alloc_size);
		if(cols_temp == NULL){
			fprintf(stderr, "Could not allocate more space for columns\n");
			return -1;
		}
		graph->cols = cols_temp;
	}

	unsigned int largest_vert = MAX(v1, v2);
	if(largest_vert >= graph->num_rows){
		// Allocate one extra space, which is just equal to current number of vertices
		unsigned int *rows_temp = realloc(graph->rows, (largest_vert+2) * sizeof(*graph->rows));
		if(rows_temp == NULL){
			fprintf(stderr, "Could not allocate more space for rows\n");
			return -1;
		}
		if(graph->rows == NULL) for(size_t i = 0;i < largest_vert+2;i++) rows_temp[i] = 0;
		graph->rows = rows_temp;
		for(size_t i = graph->num_rows;i <= largest_vert;i++){
			//printf("Extending");
			graph->rows[i] = graph->rows[graph->num_rows];
		}
		graph->num_rows = largest_vert+1;
		graph->rows[largest_vert+1] = graph->num_edges;
	}

	// Find index to insert
	size_t colIdx = graph->rows[v1];
	size_t nextIdx = graph->rows[v1+1];
	size_t i = colIdx;
	while(i < nextIdx){
		if(graph->cols[i] > v2) break;
		else if(graph->cols[i++] == v2) return -1; // Don't allow duplicates
	}

	// If the vertex does not exist, increase the vertex count
	size_t row_size = nextIdx - colIdx;
	if(!row_size) graph->num_verts++;
	//printf("Inserting at index %u\n", i);

	//shiftUpf(graph->vals, graph->num_edges, i);
	memmove(graph->vals + i + 1, graph->vals + i, (graph->num_edges - i) * sizeof(*graph->vals));
	//shiftUp(graph->cols, graph->num_edges, i);
	memmove(graph->cols + i + 1, graph->cols + i, (graph->num_edges - i) * sizeof(*graph->cols));

	graph->vals[i] = weight;
	graph->cols[i] = v2;
	graph->num_edges++;

	//printf("NextIdx: %u\tv1: %u\n", nextIdx, v1);
	for(i = v1+1;i <= graph->num_rows;++i){
		++graph->rows[i];
	}
	if(graph->rows[graph->num_rows] != graph->num_edges){
		printf("Weird case!\n");
	}
	//graph->rows[graph->num_rows] = graph->num_edges;

	return 0;
}

/**	Remove the given arc from the graph. Does not shrink arrays.
	Return -1 for error, 1 if arc doesn't exist, 0 if successful.
**/
int removeArc(struct csr_graph *graph, unsigned int v1, unsigned int v2){
	if(graph == NULL) return -1;

	// Check if either vertex is out of the graph
	unsigned int largest_vert = MAX(v1, v2);
	if(largest_vert >= graph->num_rows){
		return 1;
	}

	size_t row_size = graph->rows[v1+1] - graph->rows[v1];
	if(row_size == 0){
		return 1;
	}

	// Find target vertex
	// TODO: Use binary search
	for(unsigned int i = 0;i < row_size;++i){
		if(graph->cols[i] == v2){
			// Shift cols and vals down by 1, decrement num_edges
			memmove(graph->vals + i, graph->vals + i + 1, (graph->num_edges - i) * sizeof(*graph->vals));
			memmove(graph->cols + i, graph->cols + i + 1, (graph->num_edges - i) * sizeof(*graph->cols));
			--graph->num_edges;

			unsigned int j = v1;
			unsigned int *val = graph->rows + j;
			while(++j <= graph->num_rows){
				//printf("Val (%p): %u\n", val, *val);
				--(*(++val));
			}

			break;
		}
	}

	// If all arcs for row/vertex removed, eliminate vertex
	if(--row_size == 0) graph->num_verts--;

	return 0;
}

/**	Returns a pointer to the location in the graph array where the neighbors of
	 v1 start. The ret_size parameter will be set to the count of v1 neighbors.
	If an invalid parameter is given, ret_size will remain untouched, and the
	 function will return NULL.
**/
unsigned int * getNeighbors(struct csr_graph const *graph, unsigned int v1, size_t *ret_size){
	if(graph == NULL) return NULL;

	if(v1 >= graph->num_rows){
		return NULL;
	}

	if(ret_size != NULL) *ret_size = graph->rows[v1+1] - graph->rows[v1];
	return &graph->cols[graph->rows[v1]];
}

/**	Returns a pointer to the location in the graph array where the weights of
	 v1 neighbors start. The ret_size parameter will be set to the count of weights.
	If an invalid parameter is given, ret_size will remain untouched, and the
	 function will return NULL.
**/
double * getWeights(struct csr_graph const *graph, unsigned int v1, size_t *ret_size){
	if(graph == NULL) return NULL;

	if(v1 >= graph->num_rows){
		return NULL;
	}

	if(ret_size != NULL) *ret_size = graph->rows[v1+1] - graph->rows[v1];
	return &graph->vals[graph->rows[v1]];
}

/**	Print out all 3 arrays in the graph
**/
void dumpGraph(struct csr_graph *graph){
	if(graph == NULL) return;

	for(size_t i = 0;i < graph->num_edges;i++){
		printf("%.1f, ", graph->vals[i]);
	}
	printf("\n");
	for(size_t i = 0;i < graph->num_edges;i++){
		printf("%u, ", graph->cols[i]);
	}
	printf("\n");
	for(size_t i = 0;i <= graph->num_rows;i++){
		printf("%u, ", graph->rows[i]);
	}
	printf("\n");
}
